File: daily/cpp/2908_min_sum_of_mountain_tripliets_I.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  int minimumSum(vector<int>& nums) {
    int n = nums.size();
    if (n < 3) return -1;

    vector<int> suf(n);  // suffix min
    suf[n - 1] = nums[n - 1];
    for (int i = n - 2; i > 1; i--) {
      suf[i] = min(suf[i + 1], nums[i]);
    }

    int ans = INT_MAX;
    int pre = nums[0];  // min prefix

    for (int j = 1; j < n - 1; j++) {
      if (pre < nums[j] && nums[j] > suf[j + 1]) {
        ans = min(ans, pre + nums[j] + suf[j + 1]);
      }
      pre = min(pre, nums[j]);
    }
    return ans == INT_MAX ? -1 : ans;
  }
};

File: daily/rust/2908_min_sum_of_mountain_tripliets_I.rs
impl Solution {
    pub fn minimum_sum(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let mut nums = nums;
        if n < 3 {
            return -1;
        }

        let mut suf = vec![0; n];
        suf[n - 1] = nums[n - 1];
        for i in (2..(n - 1)).rev() {
            // suf[i] = std::cmp::min(suf[i+1], nums[i]);
            suf[i] = suf[i + 1].min(nums[i]);
        }

        let mut ans = i32::MAX;
        let mut pre = nums[0]; // min prefix

        for j in 1..(n - 1) {
            if (pre < nums[j] && nums[j] > suf[j + 1]) {
                ans = ans.min(pre + nums[j] + suf[j + 1]);
            }
            pre = pre.min(nums[j]);
        }

        if ans == i32::MAX {
            -1
        } else {
            ans
        }
    }
}
